### PR TITLE
fix(ui): Update numeric Input components to pass NaN where appropriat…

### PR DESCRIPTION
**Pull request by TCL735**
Closes https://github.com/influxdata/influxdb/issues/15024

Updates numeric Inputs to deliberately pass NaN when user input is completely deleted, as clockface will handle this appropriately, see https://github.com/influxdata/clockface/pull/373

Fixes edge cases with bin sizes in Heat Maps and decimal places in Single Stat.

### CODE REVIEW

File: ui/src/shared/utils/convertUserInput.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import React from 'react'
+
+export const convertUserInputToNumOrNaN = (
+  e: React.ChangeEvent<HTMLInputElement>
+) => (e.target.value === '' ? NaN : Number(e.target.value))
```
Comment (Asalem1) [inline ui/src/shared/utils/convertUserInput.ts]: 🤔 this ternary might cause us issues if `value` is `undefined`. Is there a possibility that the event won't have a `target` or `value` property?

File: ui/src/shared/utils/convertUserInput.test.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import {convertUserInputToNumOrNaN} from 'src/shared/utils/convertUserInput'
```
Comment (Asalem1) [inline ui/src/shared/utils/convertUserInput.test.ts]: Great job on the tests!

**Comment by TCL735** [comment]
> The PR looks great! I think it could be effective to write up a cypress e2e test to see if the app functions as expected as well.

Good idea. I will attempt to write an e2e for this and will update status here.

File: ui/src/shared/utils/convertUserInput.ts (hunk 1 above)
Comment (TCL735) [reply ui/src/shared/utils/convertUserInput.ts]: I think React.ChangeEvents always have a target. As long as we have target, value can be undefined. When value is undefined, the ternary will evaluate to false and Number(undefined) will be NaN.

Comment (Asalem1) [reply ui/src/shared/utils/convertUserInput.ts]: gotcha. Sweet!

File: ui/src/shared/utils/convertUserInput.test.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  })
+
+  it('should convert an empty string to NaN', () => {
+    value = ''
+    event.target.value = value
+
+    result = convertUserInputToNumOrNaN(event)
+    expect(Number.isNaN(result)).toEqual(true)
+  })
+})
```
Comment (hoorayimhelping) [inline ui/src/shared/utils/convertUserInput.test.ts]: nice tests!

what do you think about adding tests for some falsey values like `null`, `undefined`, and `false` since it seems like from below, the code handles those cases?

File: ui/src/timeMachine/components/view_options/DecimalPlaces.tsx [hunk 4]
```diff
@@ -65,29 +70,23 @@ class DecimalPlacesOption extends PureComponent<Props, State> {
   }
 
   public handleSetValue = (e: ChangeEvent<HTMLInputElement>): void => {
-    const value = Number(e.target.value) || 0
+    const value = convertUserInputToNumOrNaN(e)
     const {digits, onDecimalPlacesChange} = this.props
 
-    if (value === null) {
-      onDecimalPlacesChange({digits, isEnforced: false})
+    if (value === value && value >= MIN_DECIMAL_PLACES) {
```
Comment (hoorayimhelping) [inline ui/src/timeMachine/components/view_options/DecimalPlaces.tsx]: do you think the logic here is worth testing?

Comment (TCL735) [reply ui/src/timeMachine/components/view_options/DecimalPlaces.tsx]: Definitely worth testing. I will cover this with e2e tests, which will be pushed once clockface releases 1.0.5

Comment (TCL735) [reply ui/src/timeMachine/components/view_options/DecimalPlaces.tsx]: So, it turns out that we can adjust the number of decimals, but we will not be able to see the an actual number with decimal places being updated to the correct decimal place in the e2e tests. But we do have tests to validate that the input field is in default or error status.

File: ui/src/shared/utils/convertUserInput.test.ts (hunk 3 above)
Comment (TCL735) [reply ui/src/shared/utils/convertUserInput.test.ts]: Yes, will do.

File: ui/cypress/e2e/explorer.test.ts [hunk 5]
```diff
@@ -34,6 +34,150 @@ describe('DataExplorer', () => {
...
+            .type('{backspace}3')
+            .invoke('val')
+            .should('equal', '3')
+            .getByTestID('input-field--error')
+            .should('have.length', 0)
+        })
+      })
+    })
+  })
+
```
Comment (hoorayimhelping) [inline ui/cypress/e2e/explorer.test.ts]: 👏

File: ui/cypress/e2e/notificationRules.test.ts [hunk 6]
```diff
@@ -47,6 +47,234 @@ describe('NotificationRules', () => {
...
+                          })
+                      })
+                  })
+                })
+              })
+          })
+      })
+    })
+  })
+
```
Comment (hoorayimhelping) [inline ui/cypress/e2e/notificationRules.test.ts]: ❤️
